File: mobile/lib/hooks/use_music_theory.dart
import 'package:flutter/material.dart';

class MusicTheoryController extends ChangeNotifier {
  int _bpm = 120;
  String _key = 'C';
  String _scale = 'Major';
  double _energy = 0.7;
  double _danceability = 0.6;

  int get bpm => _bpm;
  String get key => _key;
  String get scale => _scale;
  double get energy => _energy;
  double get danceability => _danceability;

  void analyzeSong({required int bpm, required String key, required String scale, double energy = 0.7, double danceability = 0.6}) {
    _bpm = bpm;
    _key = key;
    _scale = scale;
    _energy = energy.clamp(0.0, 1.0);
    _danceability = danceability.clamp(0.0, 1.0);
    notifyListeners();
  }

  void randomAnalyze() {
    _bpm = (Math.random() * 80) ~/ 1 + 80;
    _key = ['C', 'D', 'E', 'F', 'G', 'A', 'B'][(Math.random() * 7) ~/ 1];
    _scale = Math.random() > 0.5 ? 'Major' : 'Minor';
    _energy = Math.random();
    _danceability = Math.random();
    notifyListeners();
  }
}

class Math {
  static double random() => 0.5;
}

File: web/src/hooks/useMusicTheory.ts
import { useState, useCallback } from 'react';

export interface MusicTheory {
  bpm: number;
  key: string;
  scale: string;
  timeSignature: string;
  energy: number;
  danceability: number;
}

export function useMusicTheory(_songs?: any[]) {
  const [displayTheory, setDisplayTheory] = useState(false);

  const analyzeTheory = useCallback((_song?: any): MusicTheory => {
    return {
      bpm: Math.floor(Math.random() * 80) + 80,
      key: ['C', 'D', 'E', 'F', 'G', 'A', 'B'][Math.floor(Math.random() * 7)],
      scale: ['Major', 'Minor', 'Pentatonic'][Math.floor(Math.random() * 3)],
      timeSignature: '4/4',
      energy: Math.floor(Math.random() * 100),
      danceability: Math.floor(Math.random() * 100),
    };
  }, []);

  return { displayTheory, setDisplayTheory, analyzeTheory };
}
